snuba/datasets/__init__.py

```diff
@@ -59,13 +64,35 @@ def column_expr(self, column_name, body):
         """
         Return an expression for the column name. Handle special column aliases
         that evaluate to something else.
-
         """
-        return column_name
+        return escape_col(column_name)
 
     def get_bulk_loader(self, source, dest_table):
         """
         Returns the instance of the bulk loader to populate the dataset from an
         external source when present.
         """
         raise NotImplementedError
+
+
+class TimeSeriesDataset(Dataset):
+    def __init__(self, *args, time_group_columns: Mapping[str, str], **kwargs):
+        super().__init__(*args, **kwargs)
+        # Convenience columns that evaluate to a bucketed time. The bucketing
+        # depends on the granularity parameter.
+        self.__time_group_columns = time_group_columns
+
+    def __time_expr(self, column_name: str, granularity: int) -> str:
+        real_column = self.__time_group_columns[column_name]
+        template = {
+            3600: 'toStartOfHour({column})',
+            60: 'toStartOfMinute({column})',
+            86400: 'toDate({column})',
+        }.get(granularity, 'toDateTime(intDiv(toUInt32({column}), {granularity}) * {granularity})')
```

Comment:
Would it make sense to limit the minimum allowed granularities per dataset ? Like not allowing someone to group by transactions by 2 5 seconds timeframes ?

Reply:
No strong opinion really, I don't think it really makes a performance difference but I also don't think there's a reason anybody should need/want extreme flexibility either here.